src/main/java/org/wildfly/security/auth/server/SecurityDomain.java

```diff
@@ -289,6 +289,12 @@ public SecurityIdentity authenticate(Principal principal, Evidence evidence) thr
         if (principal != null) serverAuthenticationContext.setAuthenticationPrincipal(principal);
         if (serverAuthenticationContext.verifyEvidence(evidence)) {
             if (serverAuthenticationContext.authorize()) {
+                Credential credential = evidence.asCredential();
```

Comment:
I don't like making this generic because clear passwords are the only case where these really align.  This is more like a mechanism (like PLAIN) where the credential is acquired, and if it's clear, then it could be forwarded.  I'd rather see a special check here to see if the evidence is a org.wildfly.security.evidence.PasswordGuessEvidence, and if so, a PasswordCredential is explicitly created.

If we have a generic asCredential() method, I worry that we'll end up with two parallel class hierarchies with all the same types in them.